bokehjs/src/lib/models/sources/ajax_data_source.ts

```diff
@@ -58,13 +60,16 @@ export class AjaxDataSource extends WebDataSource {
     }
   }
 
-  get_data(mode: UpdateMode, max_size: number | null = null, _if_modified: boolean = false): void {
+  get_data(mode: UpdateMode, max_size: number | null = null, if_modified: boolean = false): void {
     const xhr = this.prepare_request()
 
-    // TODO: if_modified
     xhr.addEventListener("load", () => this.do_load(xhr, mode, max_size ?? undefined))
     xhr.addEventListener("error", () => this.do_error(xhr))
 
+    if (if_modified && this.last_fetch_time != null) {
+      xhr.setRequestHeader("If-Modified-Since", format_date_rfc7231(this.last_fetch_time))
```

Comment:
```suggestion
      xhr.setRequestHeader("If-Modified-Since", this.last_fetch_time.toUTCString())
```

If you apply this change, you can remove your `format_date_rfc7231()` function. I also think that your tests could be reduced.